### ISSUE TITLE
Display on external monitor.

### ISSUE THREAD

**Issue by rokejulianlockhart** [axel358/smartdock#35]
When I connect an external monitor to my Fairphone 4 (the only time I'd want to use this on a smartphone, really) Smart Dock doesn't display. Does it support external displays? If so, how can I enable this?

I'd ask this as a Q&A first, but https://github.com/axel358/smartdock/issues/34.

**Comment by gaussandhisgun** [comment]
Advanced/Prefer last attached display should do what you need.

**Comment by SecularSteve** [comment]
> Advanced/Prefer last attached display should do what you need.

Tried it on my Sony Xperia 1 II, and it's still not working.
Even restarted the phone and everything.
It just remains on my phone screen no matter what I do.

You need a manual "screen detect mode" that lets you run the dock on an external screen only when detected/connected with a screen dim Desktop Mode on your phone. Something like what Taskbar has done:

https://github.com/farmerbb/Taskbar

But that's just my feature suggestion. Mainly because this seams like a good alternative to the latter which is still too buggy, and hasn't been updated in 2 years.

**Comment by nakele** [comment]
Looking for same feature, ability to display only on external screen

**Comment by MastaG** [comment]
I see you're using this as the main issue now.
There is a similar application called Taskbar.
It basically works like this.

In the Taskbar application settings, there is a desktop section where you can select your default homescreen launcher.
This is chain-loaded when there is no external screen connected, e.g. CrDroid Home, Pixel Home, ONE UI whatever.

Then in the general settings of your Android device you select Taskbar as the default homescreen launcher.

So what actually happens, is that your phone will always open Taskbar when you press the Home button.
But if there is an external screen connected, Taskbar will display itself (the actual task bar..).
If there is no external screen, it will just chain-load the homescreen you've selected in the Taskbar application settings.
You can find the source code here: https://github.com/farmerbb/Taskbar

**Comment by RokeJulianLockhart** [comment]
> https://github.com/axel358/smartdock/issues/35#issuecomment-2495651546

@MastaG, per https://github.com/farmerbb/Taskbar/issues/281#issuecomment-2108878920, this might not be useful to anyone with a modern smartphone anymore.

**Comment by MastaG** [comment]
> I see you're using this as the main issue now. There is a similar application called Taskbar. It basically works like this.
> 
> In the Taskbar application settings, there is a desktop section where you can select your default homescreen launcher. This is chain-loaded when there is no external screen connected, e.g. CrDroid Home, Pixel Home, ONE UI whatever.
> 
> Then in the general settings of your Android device you select Taskbar as the default homescreen launcher.
> 
> So what actually happens, is that your phone will always open Taskbar when you press the Home button. But if there is an external screen connected, Taskbar will display itself (the actual task bar..). If there is no external screen, it will just chain-load the homescreen you've selected in the Taskbar application settings. You can find the source code here: https://github.com/farmerbb/Taskbar

Yes even on Android 15 (CrDroid running on my old OnePlus 7T Pro).
Taskbar will load the CrDroid Home launcher on my phone screen and when I plug it into my HDMI dock, Taskbar will become the home launcher.


https://github.com/user-attachments/assets/ab9e62ca-8a14-4ac7-9584-8e679e6dc30d

**Comment by dkannegi** [comment]
Smartdock even when relocated to a SCRCPY virtual display by toggling it's accessibility service off/on, will still launch apps on the primary display on my Android 12 device. Another launcher, Hyperdroid, doesn't exhibit this cross-display launching behavior (although it lacks freeform launching like SmartDock/Taskbar have).

Smartdock: SCRCPY version tested was current stable (3.1) with the --new-display=1920x1080/180 --start=cu.axel.smartdock options.

**Comment by Xaero252** [comment]
I'd also like to see this - and possibly with the added ability to toggle the navigation bar on/off when connecting to the external display.

Currently, with Taskbar this isn't possible either - so we end up with an awkward gap on the external display which is where the android navigation bar would be - this even shows when you are in gesture navigation mode, unfortunately. It would be nice to fix both problems at the same time.

Some maybe useful insight, when connecting an external display, some useful information is provided in dumpsys json
Mostly, the device product info, and the External Touch and External Display functionality exposed here. This info would allow users to specify profiles based on devices. An example use case would be this NexDock device. It has both a touchscreen and a keyboard and trackpad. If the device is operated as a "lapdock" all three are enabled and detected by android. This would be the "desktop mode" where we'd want to use SmartDock. If the keyboard/mouse are not detected, but the touch screen is - we'd want to fall back to regular tablet mode (default screen mirroring, rather than desktop/external display mode)

I've been working on doing all of this via tasker profiles etc - but it would be nice if there was a purpose-built solution.

```json
{
  "DisplayDeviceInfo": {
    "name": "HDMI Screen",
    "uniqueId": "local:4640585011082659725",
    "resolution": "1920 x 1080",
    "modeId": 224,
    "renderFrameRate": 60.0,
    "defaultModeId": 224,
    "userPreferredModeId": -1,
    "supportedModes": [
      {
        "id": 223,
        "width": 1920,
        "height": 1080,
        "fps": 60.04206,
        "vsync": 60.04206,
        "synthetic": false,
        "alternativeRefreshRates": [23.97608, 24.0, 25.0, 29.9701, 30.0, 50.0, 59.9402, 60.0],
        "supportedHdrTypes": []
      },
      ...
    ],
    "colorMode": 0,
    "supportedColorModes": [0, 7, 9],
    "hdrCapabilities": {
      "mSupportedHdrTypes": [],
      "mMaxLuminance": 500.0,
      "mMaxAverageLuminance": 500.0,
      "mMinLuminance": 0.0
    },
    "isForceSdr": false,
    "allmSupported": false,
    "gameContentTypeSupported": false,
    "density": 213,
    "dpi": "168.165 x 161.364",
    "appVsyncOff": 6233332,
    "presDeadline": 11500001,
    "touch": "EXTERNAL",
    "rotation": 0,
    "type": "EXTERNAL",
    "address": {
      "port": 141,
      "model": "0x4066ab7b0f9587"
    },
    "deviceProductInfo": {
      "name": "NexDock",
      "manufacturerPnpId": "YUK",
      "productId": 16628,
      "modelYear": null,
      "manufactureDate": {
        "week": 44,
        "year": 2019
      },
      "connectionToSinkType": 2
    },
    "state": "ON",
    "committedState": "ON",
    "frameRateOverride": "",
    "brightnessMinimum": 0.0,
    "brightnessMaximum": 1.0,
    "brightnessDefault": 0.5,
    "hdrSdrRatio": null,
    "flags": [
      "FLAG_ALLOWED_TO_BE_DEFAULT_DISPLAY",
      "FLAG_PRESENTATION",
      "FLAG_TRUSTED"
    ],
    "installOrientation": 0,
    "displayShape": {
      "spec": 1756299619,
      "displayWidth": 1920,
      "displayHeight": 1080,
      "physicalPixelDisplaySizeRatio": 1.0,
      "rotation": 0,
      "offsetX": 0,
      "offsetY": 0,
      "scale": 1.0
    }
  }
}```

**Comment by Username1-a** [comment]
I tried that build on Android 15 (PixelOS Custom ROM) with Scrcpy v3.3.3 virtual/second display and "--start-app=cu.axel.smartdock" parameter (it only launches Smartdock settings windows from what I can see), here's some of the issues I had (Smartdock has all of the requested permissions granted):

After doing a clean installation and trying that actions build, enabling "Prefer secondary displays" and after enabling it's Accessibility permission, 
it immediately launches (with taskbar showing up) on a regular main phone screen and the only way to disable it is manually disabling it's Accessibility permission, 
but then it won't auto launch with secondary display connected (I have to manually retoggle Accessibility permission (disable then enable it) in secondary display mode for Smartdock taskbar to appear (like in previous stable builds pretty much) and if I didn't manually disable Accessibility permission before disconnecting the secondary display, taskbar will then appear on the main phone screen 


(Some sidenote issue, if Smartdock launched on secondary screen and phone screen didn't automatically turned off, when trying to manually turn the phone screen off by pressing the power button (so it wouldn't burn in further), Smartdock taskbar immediately disappears and then it immediately reappears after turning the phone screen back on (Wonder if something can so it would stay with the phone screen turned off (Tried enabling Device Admin persmission for Smartdock too)


TLDR: In my case I still need to manually enable or disable Accessibility option when entering/exiting secondary display for taskbar to appear on secondary and dissapear on the main phone screen
